app/common/renderer/components/Session/AttachToSession.jsx

```diff
@@ -1,44 +1,68 @@
-    importantCaps.push(caps.automationName);
-  }
-  return importantCaps.join(', ').trim();
+// Format the session string for standard Appium server connections
+const formatStandardCaps = (caps) => {
+  let returnedCaps = [];
+  // add deviceName OR avd OR udid
+  returnedCaps.push(caps.deviceName || caps.avd || caps.udid);
+  // add platformName and platformVersion
+  const platformInfo = caps.platformVersion
+    ? `${caps.platformName} ${caps.platformVersion}`
+    : caps.platformName;
+  returnedCaps.push(platformInfo);
+  // add automationName
+  returnedCaps.push(caps.automationName);
+  // add app OR bundleId OR appPackage
+  returnedCaps.push(caps.app || caps.bundleId || caps.appPackage);
+  return returnedCaps;
 };
 
-const formatCapsBrowserstack = (caps) => {
-  let importantCaps = formatCaps(caps).split(', ');
-  if (caps.sessionName) {
-    importantCaps.push(caps.sessionName);
+// Format the session string for cloud service provider connections
+const formatVendorCaps = (initialCaps, serverType) => {
+  let returnedCaps = [];
+  // LambdaTest may use a slightly different format
+  // and package all caps in the 'capabilities' property
+  const caps =
+    serverType === ServerTypes.lambdatest && 'capabilities' in initialCaps
+      ? initialCaps.capabilities
+      : initialCaps;
+  // add sessionName
+  returnedCaps.push(caps.sessionName);
+  // add deviceName OR avd OR udid
+  const deviceName =
+    serverType === ServerTypes.lambdatest && 'desired' in caps
+      ? caps.desired.deviceName
+      : caps.deviceName;
+  returnedCaps.push(deviceName || caps.avd || caps.udid);
+  // add platformName and platformVersion
+  if (caps.platformName) {
+    const platformInfo = caps.platformVersion
+      ? `${caps.platformName} ${caps.platformVersion}`
+      : caps.platformName;
+    returnedCaps.push(platformInfo);
   }
-  return importantCaps.join(', ').trim();
-};
-
-const formatCapsLambdaTest = (caps) => {
-  if (caps.hasOwnProperty.call(caps, 'capabilities')) {
-    caps = caps.capabilities;
-  }
-  const deviceName = caps.desired ? caps.desired.deviceName : caps.deviceName;
-  const importantCaps = [deviceName, caps.platformName, caps.platformVersion];
-  return importantCaps.join(', ').trim();
+  // add automationName
+  returnedCaps.push(caps.automationName);
+  // add app OR bundleId OR appPackage
+  returnedCaps.push(caps.app || caps.bundleId || caps.appPackage);
+  return returnedCaps;
 };
 
 const getSessionInfo = (session, serverType) => {
-  switch (serverType) {
-    case ServerTypes.browserstack:
-      return `${session.id} — ${formatCapsBrowserstack(session.capabilities)}`;
-    case ServerTypes.lambdatest:
-      return `${session.id} - ${formatCapsLambdaTest(session.capabilities)}`;
-    default:
-      return `${session.id} — ${formatCaps(session.capabilities)}`;
-  }
+  const formattedCaps =
+    serverType in [ServerTypes.remote, ServerTypes.local]
+      ? formatStandardCaps(session.capabilities)
```

Comment:
this implementation looks better to me, but I believe there is still a way to improve the code.
Lets just walk it step-by-step.
Our input is a capabilities object, whose structure might differ from one vendor to another. We'd like to transform this object into a session description string with a predefined format, where each value is separated with a slash character.
This means we need one function that is going to transform our caps object into the description string.
We know which values we'd like to put into our resulting string. We also know that a way to fetch each value may different from one provider to another.
Lets apply a bit of OOP approach to get what we need and also avoid unneeded code repetition:

```js
class DefaultDescriptionTransformer:
  constructor(caps) {
    this._caps = caps;
  }

  _fetchSessionId() {
      ....
  }

  _fetchDeviceName() {
     ....
  }

  _ fetchApp() {
    ...
  }

  transform() {
    const suffixItems = [
      this._fetchDeviceName(),
      this._fetchApp(),
     ....
    ];
    return `${this. _fetchSessionId() - _.compact(suffixItems).join(' / ')}`;
  }

class LambdaTestDescriptionTransformer(DefaultDescriptionTransformer) {
   // Only override different implementations specific to LT
   _fetchDeviceName() {
    ....
   }
}

class BrowserStackDescriptionTransformer(DefaultDescriptionTransformer) {
   // Only override different implementations specific to BS
}

.....

function getDescriptionTransformer(caps) {
  switch (serverType) {
    case ServerTypes.browserstack:
      return new BrowserStackDescriptionTransformer(caps);
    case ServerTypes.lambdatest:
      return new LambdaTestDescriptionTransformer(caps);
    default:
      return new DefaultDescriptionTransformer(caps);
}

...

const description = getDescriptionTransformer(caps).transform();

```

Reply:
Thanks for the detailed example - I've rewritten the code to follow this approach. I haven't tested it yet though (will only be able to on Monday).

Reply:
I believe it would be super easy to write simple unit test(s) for this implementation

Reply:
Added 3 unit tests.